src/lib.rs

```diff
@@ -38,13 +38,37 @@ struct Package {
 }
 
 /// Get the repository name from `Cargo.toml`.
+/// Fallbacks to directory name if `Cargo.toml` does not exists.
 pub fn read_repo(dir: &str) -> ::Result<String> {
   let mut dir = PathBuf::from(dir);
   dir.push("Cargo.toml");
-  let cargo_toml = fs::read_to_string(dir).context(::ErrorKind::Other)?;
-  let config: Config =
-    toml::from_str(&cargo_toml).context(::ErrorKind::Other)?;
-  Ok(config.package.repository)
+
+  if dir.as_path().exists() {
+    let cargo_toml = fs::read_to_string(dir).context(::ErrorKind::Other)?;
+    let config: Config =
+      toml::from_str(&cargo_toml).context(::ErrorKind::Other)?;
+    Ok(config.package.repository)
+  } else {
+    dir.pop();
+    match dir.file_name() {
+      Some(file_name) => {
+        let config: Config = Config {
+          package: Package {
+            repository: file_name.to_str().unwrap().to_string(),
```

Comment:
`.unwrap()` here means this function will crash on error. Instead returning errors using `?` is preferred.

Replace `.unwrap()` with `.context(::ErrorKind::Other)?`

Reply:
I have one question to this. How can a `context()` be applied to a method that returns an `Option`, since `to_str()` returns an `Option` here ? I tried to convert it via `ok_or()` to a `Result` but had no success and getting a 

```
error[E0599]: no method named `context` found for type `std::result::Result<&std::ffi::OsStr, &str>` in the current scope
  --> src/lib.rs:61:44
   |
61 |   let filename = dir.file_name().ok_or("").context(::ErrorKind::Other)?;
   |                                            ^^^^^^^
   |
   = note: the method `context` exists but the following trait bounds were not satisfied:
           `std::result::Result<&std::ffi::OsStr, &str> : failure::ResultExt<_, _>`
   = help: items from traits can only be used if the trait is in scope
```

Reply:
The way to do this is by using `.ok_or_else()`:

```rust
let filename = dir.file_name().ok_or_else(|| ::ErrorKind::Other)?;
```
This maps the error case to be `Err(::ErrorKind::Other)` which can be safely returned from a function using the try operator. This works because our `Error` struct knows how to convert [from `ErrorKind` to `Error`](https://github.com/yoshuawuyts/changelog/blob/master/src/error.rs#L74-L79).

There's definitely a bit of magic type casting going on here. Hopefully this will get a bit better in the future once the [`Try` trait](https://doc.rust-lang.org/nightly/core/ops/trait.Try.html) stabilizes, and we could use `.context()?` directly on an `Option`. Until then I just think of this as a bit of boilerplate to work around that.

I hope this makes sense!

Reply:
Thanks a lot for the explanation, very helpful. Always wondered how to do it the right way🎉